### PR TITLE
RCHAIN-3250 Refactor CasperPacketHandler

### PR DESCRIPTION
**Pull request by rabbitonweb**
## Overview
Refactor CasperPacketHandler to better understand what's going on.

### JIRA ticket:
RCHAIN-3250

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 1]
```diff
@@ -30,127 +30,164 @@ import monix.execution.Scheduler
...
-          _                   <- MultiParentCasperRef[F].set(casper)
-          _                   <- Log[F].info("Making a transition to ApprovedBlockRecievedHandler state.")
-          abh                 = new ApprovedBlockReceivedHandler[F](casper, approvedBlock)
-          validator           <- Ref.of[F, CasperPacketHandlerInternal[F]](abh)
-          casperPacketHandler = new CasperPacketHandlerImpl[F](validator)
-        } yield casperPacketHandler
+        val msg    = "Found ApprovedBlock in storage, reconnecting to existing network"
+        val action = connectToExistingNetwork[F](approvedBlock, init)
+        (msg, action)
+      case None if (init.conf.approveGenesis) =>
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: Guards in pattern matching disable the compiler check for exhaustiveness so I would rather not go this way

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 2]
```diff
@@ -30,127 +30,164 @@ import monix.execution.Scheduler
...
-                        conf.shardId,
-                        conf.deployTimestamp
+        val msg    = "ApprovedBlock not found in storage, connecting to existing network"
+        val action = connectAndQueryApprovedBlock[F](init)
+        (msg, action)
+    } >>= {
+      case (msg, action) => Log[F].info(msg) >> action
+    }
+
+  def connectToExistingNetwork[F[_]: LastApprovedBlock: Metrics: BlockStore: ConnectionsCell: NodeDiscovery: TransportLayer: ErrorHandler: RPConfAsk: SafetyOracle: Sync: Concurrent: Time: Log: MultiParentCasperRef: BlockDagStorage](
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: `Concurrent` is `Sync`

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 3]
```diff
@@ -30,127 +30,164 @@ import monix.execution.Scheduler
...
+      validatorId <- ValidatorIdentity.fromConfig[F](init.conf)
+      genesis     = approvedBlock.candidate.flatMap(_.block).get
+      casper <- MultiParentCasper.hashSetCasper[F](
+                 init.runtimeManager,
+                 validatorId,
+                 genesis,
+                 init.conf.shardId
+               )
+      _                   <- MultiParentCasperRef[F].set(casper)
+      _                   <- Log[F].info("Making a transition to ApprovedBlockRecievedHandler state.")
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: It's a good moment to fix the `Recieved` typo

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 4]
```diff
@@ -155,20 +153,16 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
...
-                  init.conf.approveGenesisInterval,
-                  init.conf.shardId,
-                  init.runtimeManager,
-                  validatorId,
-                  standalone
-                )
-            ).forkAndForget.runToFuture
-            ().pure[F]
-          }
+      _ <- Concurrent[F].start(
```
Comment (lukasz-golebiewski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: :100: 
:1st_place_medal:

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala (hunk 2 above)
Comment (rabbitonweb) [reply casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: Fixed

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala (hunk 3 above)
Comment (rabbitonweb) [reply casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: hopefully no integration tests rely on it :D

**Comment by ArturGajowy** [comment]
@rabbitonweb the integration test timeout seems to be consistent. Maybe it's the 'remove atrocity' change? :)

**Comment by rabbitonweb** [comment]
@ArturGajowy what I don't get is the fact that integration tests are passing...

**Comment by rabbitonweb** [comment]
einstein

**Comment by rabbitonweb** [comment]
Build have failed on CI with "insufficient memory"

```
Picked up _JAVA_OPTIONS: -Xms2G -Xmx4G -XX:MaxMetaspaceSize=1G -Dsbt.task.timings=true -Dsbt.task.timings.on.shutdown=true -Dsbt.task.timings.threshold=2000
OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x00000006c0800000, 2147483648, 0) failed; error='Not enough space' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 2147483648 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /drone/src/github.com/rchain/step.run-unit-tests/graphz/hs_err_pid468.log
[error] Could not accept connection from test agent: class java.net.SocketException: Socket closed
[error] java.net.SocketException: Socket closed
[error] 	at java.base/java.net.PlainSocketImpl.socketAccept(Native Method)
```

I will retry.

**Comment by rabbitonweb** [comment]
Modified `Fix typo ApprovedBlockReceivedHandler and reformat`

**Comment by rabbitonweb** [comment]
`test_casper_propose_and_deploy` failed on CI, passing locally, retry

**Pull request merged**